File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> mp = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"}, ans;
    vector<string> letterCombinations(string digits) {
        if(digits == "") return {};
        combi(0, "", digits);
        return ans;
    }
    void combi(int i, string curr, string digits){
        if(i == digits.size()){
            ans.push_back(curr);
            return;
        }
        for(auto x : mp[digits[i]-'2']){
            curr.push_back(x);
            combi(i+1, curr, digits);
            curr.pop_back();
        }
    }
};

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    String[] mp = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    List<String> ans = new ArrayList<String>();
    public List<String> letterCombinations(String digits) {
        if(digits.isEmpty()) return ans;
        helper(0, new StringBuilder(), digits);
        return ans;
    }
    void helper(int i, StringBuilder combi, String digits){
        if(i == digits.length()){
            ans.add(combi.toString());
            return;
        }
        for(char x : mp[digits.charAt(i)-'2'].toCharArray()){
            combi.append(x);
            helper(i+1, combi, digits);
            combi.deleteCharAt(combi.length()-1);
        }
    }
}